opm/parser/eclipse/Parser/Parser.cpp

```diff
@@ -42,60 +42,194 @@
+
+        return std::find_if_not( rbegin, rend, RawConsts::is_separator ).base();
+    }
+
+    static inline string_view trim( string_view str ) {
+        auto fst = trim_left( str.begin(), str.end() );
+        auto lst = trim_right( fst, str.end() );
+        return { fst, lst };
+    }
+
+    static inline string_view strip_slash( string_view view ) {
+        using itr = string_view::const_iterator;
+        const auto term = []( itr begin, itr end ) {
+            return std::find( begin, end, '/' );
+        };
+
+        auto begin = view.begin();
+        auto end = view.end();
+        auto slash = strip_after( begin, end, term );
+
+        /* we want to preserve terminating slashes */
+        if( slash != end ) ++slash;
+
+        return { begin, slash };
+    }
+
+    static inline bool getline( string_view& input, string_view& line ) {
+        if( input.empty() ) return false;
+
+        auto end = std::find( input.begin(), input.end(), '\n' );
+
+        line = string_view( input.begin(), end );
+        const auto mod = end == input.end() ? 0 : 1;
+        input = string_view( end + mod, input.end() );
+        return true;
+    }
+
+    /*
+     * Read the input file and remove everything that isn't interesting data,
+     * including stripping comments, removing leading/trailing whitespaces and
+     * everything after (terminating) slashes
+     */
+    static inline std::string clean( const std::string& str ) {
+        std::string dst;
+        dst.reserve( str.size() );
+
+        string_view input( str ), line;
+        while( getline( input, line ) ) {
+            line = trim( strip_slash( strip_comments( line ) ) );
+
+            //if( line.begin() == line.end() ) continue;
+
+            dst.append( line.begin(), line.end() );
+            dst.append( 1, '\n' );
+        }
+
+        return dst;
+    }
+
+    const std::string emptystr = "";
 
     struct ParserState {
+
+        struct file {
+            file( boost::filesystem::path p, const std::string& in ) :
+                path( p ), input( in )
+            {}
+
+            file( const std::string& in ) : path( "" ), input( in ) {}
+
+            boost::filesystem::path path;
+            string_view input;
+            size_t lineNR = 0;
+        };
+
         const ParseContext& parseContext;
         Deck * deck;
-        boost::filesystem::path dataFile;
+
         boost::filesystem::path rootPath;
-        /* the path map is shared between all state instances, as it has global
-         * resolution, i.e. PATHS specified in an included file can be used
-         * from the file that included it.
-         */
-        std::shared_ptr< std::map<std::string, std::string> > pathMap;
-        size_t lineNR;
-        std::shared_ptr<std::istream> inputstream;
+        std::map< std::string, std::string > pathMap;
+        std::list< std::string > input_strings;
+        std::list< file > input_files;
+        std::vector< file* > input_stack;
```

Comment:
This is the natural development for this (but I wanted to keep it out of this PR), especially since it allows us to re-enable streaming behaviour at the flip of a switch. I kept it out to keep the actual reading changes as small as possible, but I'm happy adding a new commit that fixes this.

Making the input stack a richer object is a good idea and was certainly on the horizon.